### DISCUSSION TITLE
Use docker images from GitHub Package Registry

### DISCUSSION THREAD

**Discussion by peaceiris** [community/community#25235] [Actions]
<p>I joined GitHub Actions and GitHub Package Registry. According to <a href="https://github.community/t/github-action-build-caching/16023" title="Solved: Github Action Build Caching - GitHub Community Forum">Solved: Github Action Build Caching - GitHub Community Forum</a>, we can skip docker build step for using GitHub Actions. GitHub Actions v2 can pull the docker images that are published on Docker Hub now. Can GitHub Actions v2 pull the docker images from GitHub Package Registry? Is this on the roadmap?</p>

**Comment by peaceiris** [comment]
<ul>
<li><a href="https://github.blog/2019-09-11-proxying-packages-with-github-package-registry-and-other-updates/?utm_campaign=1568215894&amp;utm_medium=social&amp;utm_source=twitter&amp;utm_content=1568215894" rel="nofollow noopener">Proxying packages with GitHub Package Registry and other updates</a></li>
</ul>
<p>&gt; Using packages in Actions</p>
<p>&gt;</p>
<p>&gt; We’re continuing to bring Actions and Package Registry closer together, starting with removing the need to use personal access tokens to access packages from Actions.</p>
<p>I got it. Thank you.</p>

**Comment by jsumners** [comment]
<p>I’m not sure how this solves the issue. Can you please show an example action wherein you are pulling a container image from a public GitHub Docker registry?</p>

**Comment by peaceiris** [comment]
<p>GitHub team is implementing it now.</p>
<p>Today (7 Oct.), we cannot use a docker image from GHPR on GitHub Actions.</p>
<pre><code>uses: docker://docker.pkg.github.com/peaceiris/actions-gh-pages/action:latest
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/1957ff6d15df923c5a68b4de278a4d4be888391e.png" data-download-href="/uploads/short-url/3CctcREgmx5HR7IE3pbNNmooTz8.png?dl=1" title="Screen Shot 2019-10-07 at 17.45.38.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/30958501/181089745-3dfdbc7a-2a5e-4c4a-b233-c3196051cfdb.png" alt="Screen Shot 2019-10-07 at 17.45.38.png" data-base62-sha1="3CctcREgmx5HR7IE3pbNNmooTz8" width="690" height="276" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/1/1957ff6d15df923c5a68b4de278a4d4be888391e_2_690x276.png, https://aws1.discourse-cdn.com/github/optimized/2X/1/1957ff6d15df923c5a68b4de278a4d4be888391e_2_1035x414.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/1/1957ff6d15df923c5a68b4de278a4d4be888391e_2_1380x552.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/1957ff6d15df923c5a68b4de278a4d4be888391e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-10-07 at 17.45.38.png</span><span class="informations">1442×578 121 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by arondeparon** [comment]
<p>We got this to work by simply using the Docker CLI to do this.</p>
<p>Here is my example workflow. The only downside to this workflow is that I cannot conditionally build the Docker container, so it is a bit slower than I want it to be, but I can still build my entire C project in a couple of minutes using this workflow.</p>
<pre><code>name: build firmware

on: [push]

jobs:

  build_docker_image:

    runs-on: ubuntu-18.04

    steps:

    - uses: actions/checkout@v1

    - name: Build and tag image
      run: docker build -t "docker.pkg.github.com/$GITHUB_REPOSITORY/firmware-builder:latest" -f docker/Dockerfile docker/

    - name: Docker login
      run: docker login docker.pkg.github.com -u $GITHUB_ACTOR -p $GITHUB_TOKEN
      env:
        GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}

    - name: Publish to GPR
      run: docker push "docker.pkg.github.com/$GITHUB_REPOSITORY/firmware-builder:latest"

  build_firmware:
    needs: build_docker_image
    runs-on: ubuntu-18.04

    steps:
    - uses: actions/checkout@v1

    - name: Docker login
      run: docker login docker.pkg.github.com -u $GITHUB_ACTOR -p $GITHUB_TOKEN
      env:
          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}

    - name: Pull Docker image
      run: docker pull "docker.pkg.github.com/$GITHUB_REPOSITORY/firmware-builder:latest"

    - name: Run image
      run: docker run -it -d --name builder -v $GITHUB_WORKSPACE:/workspace -w /workspace docker.pkg.github.com/$GITHUB_REPOSITORY/firmware-builder:latest

    - name: make version
      run: echo ::set-env name=FIRMWARE_VERSION::$(docker exec builder bash ./make_version.sh)
    - name: make
      run: docker exec builder bash -c 'make clean &amp;&amp; make'
    - name: save artifact
      run: docker exec builder bash -c 'cp upload_program.sh output/'

    - uses: actions/upload-artifact@master
      with:
        name: firmware-${{ env.FIRMWARE_VERSION }}
        path: output/
 
</code></pre>

**Comment by peaceiris** [comment]
<p>That’s good but a little complex. I hope that we can use docker images from the Package Registry with the <code>uses:</code> statement.</p>
<p>(Today Nov. 12th, we still cannot do that.)</p>

**Comment by arondeparon** [comment]
<p>Yeah, I agree. My current fix comes with a pretty big performance penalty, so it definitely is not ideal.</p>

**Comment by nightlark** [comment]
<p>Not requiring an auth token to install a published package makes the feature significantly less useful and impossible to use in the way that I wanted: a Docker GitHub Action that uses a container image from the GitHub Package Registry as the base in a <code>FROM</code> statement in the Dockerfile for my action.</p>
<p>There’s no way to build a Docker action that just works with a <code>uses</code> statement if the image is getting pulled from Package Registry, because it requires logging in. I’d say that is a deal-breaker for using Package Registry that will require me to set up an organization on Docker Hub and push images there instead, something I was hoping to avoid (everything using GitHub would have been amazing).</p>

**Comment by lagerfeuer** [comment]
<p>Has something changed here? I’m trying to use a Docker image from a different (public) repository to test my code, I’d like to use it from the registry so I don’t have to build it everytime, since it takes quite long.</p>
<p>Trying to log in first doesn’t work because for some reason “docker pull” is run before any job.</p>

**Comment by petejohanson** [comment]
<p>How is this <em>still</em> not solved? Two major “efforts” from GH that don’t actually work together?</p>

**Comment by gustavovalverde** [comment]
<p>Not being able to use images built on a previous <code>step</code> in a <code>uses</code> statement made me think that I could workaround it by pushing it to GitHub Package Registry. Seems like I was wrong.</p>
<p>I wonder how this even works with Docker Hub and Google Cloud Registry (GCR), but not with GPR. Seems like an issue with GPR and not Github Actions.</p>

**Comment by peaceiris** [comment]
<p>Today, we still cannot use a Docker image from GitHub Packages in a <code>uses:</code> statement.</p>
<pre><code class="lang-yaml">- uses: 'docker://docker.pkg.github.com/peaceiris/actions-gh-pages/action:latest'
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/fbee4200f7b277743c9c310307bc0da45b68f10c.png" data-download-href="/uploads/short-url/zWGje75lAczLhue0WS30z9ExAmg.png?dl=1" title="Screen Shot 2020-06-15 at 12.53.25" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/30958501/181089755-d99b510a-4b7c-445e-9b4e-74cbc21638b9.png" alt="Screen Shot 2020-06-15 at 12.53.25" data-base62-sha1="zWGje75lAczLhue0WS30z9ExAmg" width="690" height="213" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/f/fbee4200f7b277743c9c310307bc0da45b68f10c_2_690x213.png, https://aws1.discourse-cdn.com/github/optimized/2X/f/fbee4200f7b277743c9c310307bc0da45b68f10c_2_1035x319.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/f/fbee4200f7b277743c9c310307bc0da45b68f10c_2_1380x426.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/fbee4200f7b277743c9c310307bc0da45b68f10c_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-06-15 at 12.53.25</span><span class="informations">2150×666 130 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by robjmills** [comment]
<p>That’s only true of public images in GCR and Docker Hub though, right?</p>

**Comment by nightlark** [comment, marked as answer]
<p>The <a href="https://github.blog/2020-09-01-introducing-github-container-registry/" rel="nofollow noopener">GitHub Container Registry</a> announced today looks like their answer to this issue. It allows anonymous read access and it looks like it could provide more control for managing packages, though it’s accessed by a <code>ghcr.io/&lt;user_or_org_name&gt;</code> address.</p>
<p>The other downside is that it looks like a personal access token needs to be created to push to a GitHub org/user-level container repository so a <code>GITHUB_TOKEN</code> with GitHub repository scope isn’t enough, but at least now everything can live on GitHub instead of needing a separate Docker Hub account.</p>
<p>Some useful docs pages about GHCR:<br>
<a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/about-github-container-registry" rel="nofollow noopener">About GHCR</a><br>
<a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images" rel="nofollow noopener">Migrating to GHCR</a></p>

**Comment by 20k-ultra** [comment]
<p>I came here to figure out how to use my built image as the action but the solved answer didn’t provide that info. I believe <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-docker-hub-action" rel="noopener nofollow ugc">this</a> is the relevant documentation on how to use an action that is a built docker image.</p>
<pre><code class="lang-auto">jobs:
  my_first_job:
    steps:
      - name: My first step
        uses: docker://ghcr.io/OWNER/IMAGE_NAME
</code></pre>

**Reply by lestephane** [reply]
This simply does not work

```
jobs:
  test:
    ...
    permissions:
      packages: read
    steps:
      - name: "AWS Ping (cloudformation)"
        uses: docker://ghcr.io/ekalinin/awsping:v2.0.1
        with:
          args: -https -repeats 3 -service cloudformation -verbose 2

```
Leads to 
```
 /usr/bin/docker pull ghcr.io/ekalinin/awsping:v2.0.1
  Error response from daemon: Head "https://ghcr.io/v2/ekalinin/awsping/manifests/v2.0.1": unauthorized
  Warning: Docker pull failed with exit code 1, back off 4.839 seconds before retry.
  /usr/bin/docker pull ghcr.io/ekalinin/awsping:v2.0.1
  Error response from daemon: Head "https://ghcr.io/v2/ekalinin/awsping/manifests/v2.0.1": unauthorized
  Warning: Docker pull failed with exit code 1, back off 7.411 seconds before retry.
  /usr/bin/docker pull ghcr.io/ekalinin/awsping:v2.0.1
  Error response from daemon: Head "https://ghcr.io/v2/ekalinin/awsping/manifests/v2.0.1": unauthorized
Error: Docker pull failed with exit code 1
```
 
None of  examples in the docs show how this is done (ie they all leave the story of authorization undocumented).

- <https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-docker-hub-action>
- <https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-the-github-packages-container-registry>
- <https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-docker-public-registry-action>

There 
1) `packages: read` does not help because it only grants access to the current repo packages and containers.
2) an extra `docker login` step won't work because the container is pulled _in the setup phase of the job_ (at least that is what I saw in my testing), _before any step is executed_.
![image](https://user-images.githubusercontent.com/5460884/192977851-a7e4edf6-98ca-48a4-8bbb-40e8865757fc.png)

**Comment by anthonylouisbsb** [comment]
<p>I had a action that use the container registry:</p>
<pre><code class="lang-auto">jobs:
  test:
    container:
      image: docker.pkg.github.com/&lt;org&gt;/&lt;repo-name&gt;/&lt;image-name&gt;:version
      env:
        GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
</code></pre>
<p>Since the last week that job has some errors when it is setting up the tests environment:</p>
<pre><code class="lang-auto">  /usr/bin/docker pull docker.pkg.github.com/&lt;org-name&gt;/&lt;repo-id&gt;/&lt;image-name&gt;:version
  Error response from daemon: unauthorized
</code></pre>
<p>It is strange because there are some jobs does not have that problem. The behavior is inconsistent.</p>
<p>It was working fine, I did not change anything and started to crash. Does anyone has any idea of how to fix that problem?</p>

**Comment by airtower-luna** [comment]
<p><code>docker.pkg.github.com</code> is deprecated, see <a href="https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-docker-registry" class="inline-onebox">Working with the Docker registry - GitHub Docs</a>. Use <code>ghcr.io</code> instead as described there.</p>
